transport-native-epoll/src/main/java/io/netty/channel/epoll/AIOContext.java

```diff
@@ -62,34 +70,54 @@ public long getAddress() {
         assert !destroyed;
         assert dst.isDirect();
         assert dst.position() == 0;
+        assert file.epollEventLoop.aioContext == this;
+
+        int length = (dst.limit() & 511) == 0 ? dst.limit() : ((dst.limit() + 511) & ~511);
+        if (dst.capacity() < length) {
+            handler.failed(new RuntimeException("supplied buffer isn't long enough to handle read length " +
+                                                "alignment"), attachment);
+            return;
+        }
+
+        if ((position & 511) != 0) {
+            handler.failed(new IOException("Read position must be aligned to sector size (usually 512)"),
+                           attachment);
+            return;
+        }
+
+        IORequest<A> request = new IORequest<A>(file, dst, position, length, handler, attachment);
 
-        int length = dst.limit();
         try {
-            if (file.isDirect()) {
-                length = (dst.limit() & 511) == 0 ? dst.limit() : ((dst.limit() + 511) & ~511);
-                if (dst.capacity() < length) {
-                    throw new RuntimeException("supplied buffer isn't long enough to handle read length alignment");
+            // Avoid sending overflowing the aio context queue (EGAIN)
+            // instead buffer locally
+            if (outstanding + 1 <= maxOutstanding) {
+                long id = Native.submitAIORead(this, file.getEventFd(), file.getFd(), position, length, dst);
+
+                IORequest r = outstandingRequests.putIfAbsent(id, request);
+                ++outstanding;
```

Comment:
nit: I may be misunderstanding the way the code here's supposed to work, but shouldn't the incrementing of `outstanding` depend on the result of the `putIfAbsent` and/or `offer` calls??

Reply:
I agree, `outstanding` should be incremented only if the item was added to the map, which is why I suggested above to just use the map size.